test/unit/legacy_wrappers/mongo_client.test.js

```diff
@@ -22,6 +24,89 @@ describe('legacy_wrappers/mongo_client.js', () => {
+      }).to.not.throw();
+    });
+  });
+
+  describe('setting client metadata', () => {
+    it('does not mutate the input options', () => {
+      expect(() => {
+        new LegacyMongoClient(iLoveJs, Object.freeze({}));
+      }).to.not.throw();
+    });
+
+    it('does not mutate the input options.driverInfo', () => {
+      expect(() => {
+        new LegacyMongoClient(iLoveJs, Object.freeze({ driverInfo: Object.freeze({}) }));
+      }).to.not.throw();
+    });
+
+    describe('when driverInfo.name is provided', () => {
+      const client = new LegacyMongoClient(iLoveJs, { driverInfo: { name: 'mongoose' } });
+
+      it('should prepend mongodb-legacy to user passed driverInfo.name', () =>
+        expect(client.options.metadata).to.have.nested.property(
+          'driver.name',
+          'nodejs|mongodb-legacy|mongoose'
+        ));
+
+      it('should include version in package.json in client metadata', () =>
+        expect(client.options.metadata)
+          .to.have.property('version')
+          .that.includes(currentLegacyVersion));
+    });
+
+    describe('when driverInfo.name is provided and driverInfo.version is provided', () => {
+      const client = new LegacyMongoClient(iLoveJs, {
+        driverInfo: { name: 'mongoose', version: '99.99.99' }
+      });
+
+      it('should prepend mongodb-legacy to user passed driverInfo.name', () =>
+        expect(client.options.metadata)
+          .to.have.nested.property('driver.name')
+          .that.equals('nodejs|mongodb-legacy|mongoose'));
+
+      it('should prepend version in package.json to user driverInfo.version', () =>
+        expect(client.options.metadata)
+          .to.have.property('version')
+          .that.includes(`${currentLegacyVersion}|99.99.99`));
+    });
+
+    describe('when driverInfo.version is provided', () => {
+      const client = new LegacyMongoClient(iLoveJs, {
+        driverInfo: { version: '99.99.99' }
+      });
+
+      it('should include mongodb-legacy in client metadata', () =>
+        expect(client.options.metadata)
+          .to.have.nested.property('driver.name')
+          .that.equals('nodejs|mongodb-legacy'));
+
+      it('should prepend version in package.json to user driverInfo.version', () =>
+        expect(client.options.metadata)
+          .to.have.property('version')
+          .that.includes(`${currentLegacyVersion}|99.99.99`));
+    });
+
+    describe('when driverInfo is not provided', () => {
+      const client = new LegacyMongoClient(iLoveJs);
+
+      it('should include mongodb-legacy in client metadata', () =>
+        expect(client.options.metadata)
+          .to.have.nested.property('driver.name')
+          .that.equals('nodejs|mongodb-legacy'));
+
+      it('should include version in package.json in client metadata', () =>
+        expect(client.options.metadata)
+          .to.have.property('version')
+          .that.includes(currentLegacyVersion));
+    });
+  });
+
```

Comment:
Can we add tests that test that we do not modify the provided options object as well?  We could add an `it` block into each of the describes, i.e.

```typescript
describe('when driverInfo.version is provided', () => {
     const options = {
        driverInfo: { version: '99.99.99' }
      };
      const client = new LegacyMongoClient(iLoveJs, options);

      it('should include mongodb-legacy in client metadata', () =>
        expect(client.options.metadata)
          .to.have.nested.property('driver.name')
          .that.equals('nodejs|mongodb-legacy'));

      it('should prepend version in package.json to user driverInfo.version', () =>
        expect(client.options.metadata)
          .to.have.property('version')
          .that.includes(`${currentLegacyVersion}|99.99.99`));
    });

    it('should not modify the options object', () => {
        expect(options.to.deep.equal({ driverInfo: { version: '99.99.99' } })
    });
});
```

or something like this

Reply:
Isn't this what the frozen object tests cover?